File: projects/sheaf-chat/src/server/smoke_test.ts
import { existsSync, statSync } from "node:fs";

export const x_smokeTestModeEnvVar = "SHEAF_SMOKE_TEST_MODE";
export const x_smokeAssetRootEnvVar = "SHEAF_SMOKE_ASSET_ROOT";

// Smoke-test mode is active when the variable is set to any non-empty value
// other than `0` or `false` (case-insensitive).
export function IsSmokeTestModeActive(value: string | undefined): boolean
{
  if (value === undefined)
  {
    return false;
  }

  const normalized = value.trim().toLowerCase();
  return !(normalized.length === 0 || normalized === "0" || normalized === "false");
}

export interface ResolvedSmokeAssetRoot
{
  assetRoot: string;
  usedSmokeAssetRoot: boolean;
  warning: string | null;
}

export interface ResolveSmokeAssetRootOptions
{
  repoRoot: string;
  env?: NodeJS.ProcessEnv;
  directoryExists?: (candidate: string) => boolean;
}

function defaultDirectoryExists(candidate: string): boolean
{
  return existsSync(candidate) && statSync(candidate).isDirectory();
}

// Resolves the directory git-ignored assets (e.g. config/api_keys.json) should
// be loaded from. When smoke mode is inactive, returns repoRoot. When active
// and SHEAF_SMOKE_ASSET_ROOT names an existing directory, returns that. When
// active but the asset root is unset/empty/missing, returns repoRoot with a
// warning describing the fallback.
export function ResolveSmokeAssetRoot(
  options: ResolveSmokeAssetRootOptions,
): ResolvedSmokeAssetRoot
{
  const env = options.env ?? process.env;
  const directoryExists = options.directoryExists ?? defaultDirectoryExists;

  if (!IsSmokeTestModeActive(env[x_smokeTestModeEnvVar]))
  {
    return { assetRoot: options.repoRoot, usedSmokeAssetRoot: false, warning: null };
  }

  const rawAssetRoot = (env[x_smokeAssetRootEnvVar] ?? "").trim();

  if (rawAssetRoot.length === 0)
  {
    return {
      assetRoot: options.repoRoot,
      usedSmokeAssetRoot: false,
      warning: `smoke-test mode active but ${x_smokeAssetRootEnvVar} is unset or empty; using repo root for assets`,
    };
  }

  if (!directoryExists(rawAssetRoot))
  {
    return {
      assetRoot: options.repoRoot,
      usedSmokeAssetRoot: false,
      warning: `smoke-test mode active but ${x_smokeAssetRootEnvVar} (${rawAssetRoot}) is not an existing directory; using repo root for assets`,
    };
  }

  return { assetRoot: rawAssetRoot, usedSmokeAssetRoot: true, warning: null };
}

File: projects/quest-runner/src/quest_runner_service/smoke_test.py
"""Repo-wide smoke-test mode contract for the quest-runner service.

When ``SHEAF_SMOKE_TEST_MODE`` is active, the service resolves any git-ignored
assets it reads (e.g. ``config/api_keys.json``, ``.secrets.json``) from the
main-repo asset root given by ``SHEAF_SMOKE_ASSET_ROOT`` rather than from its own
repository root, while continuing to read tracked files from the repository
root.
"""

from __future__ import annotations

import os
from collections.abc import Callable, Mapping
from dataclasses import dataclass
from pathlib import Path

MODE_ENV_VAR = "SHEAF_SMOKE_TEST_MODE"
ASSET_ROOT_ENV_VAR = "SHEAF_SMOKE_ASSET_ROOT"


def is_smoke_test_mode_active(env: Mapping[str, str]) -> bool:
    """Smoke mode is active for any non-empty value other than ``0``/``false``."""
    raw = env.get(MODE_ENV_VAR)
    if raw is None:
        return False
    normalized = raw.strip().lower()
    return normalized not in ("", "0", "false")


@dataclass(frozen=True)
class ResolvedAssetRoot:
    """Where git-ignored assets should be read from, plus any fallback warning."""

    asset_root: Path
    used_smoke_asset_root: bool
    warning: str | None


def resolve_asset_root(
    repo_root: Path,
    env: Mapping[str, str],
    directory_exists: Callable[[Path], bool] = lambda path: path.is_dir(),
) -> ResolvedAssetRoot:
    """Resolve the directory git-ignored assets should be read from.

    - Smoke mode inactive: returns ``repo_root``.
    - Active with a valid ``SHEAF_SMOKE_ASSET_ROOT`` directory: returns it.
    - Active but the asset root is unset/empty/missing: returns ``repo_root``
      with a warning describing the fallback.
    """
    if not is_smoke_test_mode_active(env):
        return ResolvedAssetRoot(repo_root, False, None)

    raw_asset_root = (env.get(ASSET_ROOT_ENV_VAR) or "").strip()
    if not raw_asset_root:
        return ResolvedAssetRoot(
            repo_root,
            False,
            f"smoke-test mode active but {ASSET_ROOT_ENV_VAR} is unset or empty; "
            "using repo root for assets",
        )

    asset_root = Path(raw_asset_root)
    if not directory_exists(asset_root):
        return ResolvedAssetRoot(
            repo_root,
            False,
            f"smoke-test mode active but {ASSET_ROOT_ENV_VAR} ({raw_asset_root}) "
            "is not an existing directory; using repo root for assets",
        )

    return ResolvedAssetRoot(asset_root, True, None)


def resolve_asset_path(
    relative_path: str,
    repo_root: Path,
    env: Mapping[str, str] | None = None,
) -> Path:
    """Resolve a git-ignored asset's path, honoring smoke-test mode.

    Use this for files such as ``config/api_keys.json`` that live outside git;
    tracked files should continue to resolve against ``repo_root`` directly.
    """
    resolution = resolve_asset_root(repo_root, env if env is not None else os.environ)
    return resolution.asset_root / relative_path
